### PR TITLE
Key recovery improvements

### PR DESCRIPTION
**Pull request by Sil333033**
- Bugfix: Calculated nested key recheck didn't work.
- Bugfix: If sector is 0 in mfClassicGetSectorTrailerBlockBySector() it didnt return 3. 
- Added: Automatic key recovery. Try dictionary and switch to Nested recovery after.

### CODE REVIEW

File: chameleonultragui/lib/helpers/mifare_classic/general.dart [hunk 1]
```diff
@@ -156,7 +156,9 @@ MifareClassicType mfClassicGetCardTypeByBlockCount(int blockCount) {
 }
 
 int mfClassicGetSectorTrailerBlockBySector(int sector) {
-  if (sector < 32) {
+  if (sector == 0) {
```
Comment (GameTec-live) [inline chameleonultragui/lib/helpers/mifare_classic/general.dart:159]: why add that?

File: chameleonultragui/lib/l10n/app_en.arb [hunk 2]
```diff
@@ -167,6 +167,8 @@
   "dump_partial_data": "Dump partial data",
   "additional_key_dict": "Additional key dictionary",
   "check_keys_dict": "Check keys from dictionary",
+  "check_keys_autopwn": "Try automatic recovery",
```
Comment (GameTec-live) [inline chameleonultragui/lib/l10n/app_en.arb:170]: dont just call the key smthng with autopwn, when translating noone will know the context... Also; very long button name...

File: chameleonultragui/lib/gui/component/card_recovery.dart [hunk 3]
```diff
@@ -323,6 +323,49 @@ class CardRecoveryState extends State<CardRecovery> {
                   }
                 : null,
             child: Text(localizations.check_keys_dict),
+          ),
+          const SizedBox(height: 16),
```
Comment (GameTec-live) [inline chameleonultragui/lib/gui/component/card_recovery.dart:327]: maybe use 8 to fit in with the rest, but ill look at it when testing...

**Comment by GameTec-live** [comment]
image
why arent you disabling it when prng is hard?
this should be disabled when prng is hard...

File: chameleonultragui/lib/helpers/mifare_classic/general.dart (hunk 1 above)
Comment (Sil333033) [reply chameleonultragui/lib/helpers/mifare_classic/general.dart:159]: my fault, forgot to remove this. i was convinced this gave an bug to my added code

File: chameleonultragui/lib/l10n/app_en.arb (hunk 2 above)
Comment (Sil333033) [reply chameleonultragui/lib/l10n/app_en.arb:170]: i mean the button above is "Check keys from dictionary". thats a bit longer then "Try automatic recovery".

but idc, change it how you want

**Comment by Sil333033** [comment]
Why? it can still perform an dictionary attack. i have a couple of PRNG hard cards here that use FFFFFFFFFFFF, so it works fine on those cards

**Comment by GameTec-live** [comment]
fixed with #436; thanks for finding the bug.

to anyone else:
Both are very similar in speed and very card dependent so we (Sil and me) decided to not merge and just push the bugfixes...

**Pull request closed without merging**